### DISCUSSION TITLE
YOLO-6D-Pose Multi-Object 6D Pose Estimation Model

### DISCUSSION THREAD

**Discussion by tv-gc** [asus4/onnxruntime-unity-examples#11] [Q&A]
Hi!

I am trying to use the onnx files from this repo in order to obtain 6d pose information on objects

https://github.com/TexasInstruments/edgeai-yolox/blob/main/README_6d_pose.md

more especifically, the default model yolox_s_object_pose.onnx converted to Ort as requested.

![image](https://github.com/asus4/onnxruntime-unity-examples/assets/90689163/e627ed6e-d611-4785-8227-10bec16a8cca)

But trying to run the YOLOx example, i am getting this error

```
OnnxRuntimeException: [ErrorCode:RuntimeException] D:\a\_work\1\s\onnxruntime\core\framework\tensor.cc:33 onnxruntime::Tensor::CalculateTensorStorageSize shape.Size() must >=0

Microsoft.ML.OnnxRuntime.NativeApiStatus.VerifySuccess (System.IntPtr nativeStatus) (at ./Library/PackageCache/com.github.asus4.onnxruntime@0.1.12/Runtime/NativeApiStatus.shared.cs:33)
Microsoft.ML.OnnxRuntime.OrtValue.CreateAllocatedTensorValue (Microsoft.ML.OnnxRuntime.OrtAllocator allocator, Microsoft.ML.OnnxRuntime.Tensors.TensorElementType elementType, System.Int64[] shape) (at ./Library/PackageCache/com.github.asus4.onnxruntime@0.1.12/Runtime/OrtValue.shared.cs:620)
Microsoft.ML.OnnxRuntime.Unity.InferenceSessionExtension.CreateTensorOrtValue (Microsoft.ML.OnnxRuntime.NodeMetadata metadata) (at ./Library/PackageCache/com.github.asus4.onnxruntime.unity@0.1.12/Runtime/Extensions/InferenceSessionExtension.cs:55)
Microsoft.ML.OnnxRuntime.Unity.ImageInference`1[T].AllocateTensors (System.Collections.Generic.IReadOnlyDictionary`2[TKey,TValue] metadata) (at ./Library/PackageCache/com.github.asus4.onnxruntime.unity@0.1.12/Runtime/ImageInference.cs:149)
Microsoft.ML.OnnxRuntime.Unity.ImageInference`1[T]..ctor (System.Byte[] model, Microsoft.ML.OnnxRuntime.Unity.ImageInferenceOptions options) (at ./Library/PackageCache/com.github.asus4.onnxruntime.unity@0.1.12/Runtime/ImageInference.cs:64)
Microsoft.ML.OnnxRuntime.Examples.Yolox..ctor (System.Byte[] model, Microsoft.ML.OnnxRuntime.Examples.Yolox+Options options) (at Assets/Samples/Yolox/Yolox.cs:105)
YoloxSample.Start () (at Assets/Samples/Yolox/YoloxSample.cs:33)
```

Would it be possible to help me with this problem?

Cheers!!

**Comment by asus4** [comment]
Hi @tv-gc 

That error is usually caused by the dynamic shape `Gatherdetections_dim_0`.  You need to define the shape size yourself. The `ImageInference.cs` and `Yolox.cs` doesn't support the dynamic shape right now. Please allocate the OrtValue from scratch.


![Screenshot 2024-02-07 at 14 46 32](https://github.com/asus4/onnxruntime-unity-examples/assets/357497/8fea7bb3-3d4e-4bb0-be6c-a935935e17c3)

**Comment by tv-gc** [comment]
Thanks for the super quick reply, first of all :)

So, instead of letting the base class handling 

```
 // Allocate inputs/outputs
 inputs = AllocateTensors(session.InputMetadata);
 outputs = AllocateTensors(session.OutputMetadata);
```

I need to define manually instead of passing session.InputMetadata?
Or i need to create a InferenceSession from scratch?

Sorry but i don't have a lot of experience with onnxruntime and i am struggling a bit with this error.

Cheers

**Comment by asus4** [comment]
@tv-gc No worries. You probably need to copy the `ImageInference.cs` since you can't override that constructor method from the sub-class.

I will add some options to the ImageInference that accept the dynamic shape model when I have time.

**Comment by tv-gc** [comment]
Much appreciated!

**Comment by tv-gc** [comment]
I believe i managed to apply a workaround for this particular problem but now the outputs are empty.

I have copied the Inference class and tweaked the CreateTensorOrtValue a bit to avoid the -1 value and instead place a 0

```
private static OrtValue CreateTensorOrtValueDynamic(NodeMetadata metadata)
{
if (!metadata.IsTensor)
{
throw new ArgumentException("metadata must be tensor");
}

int[] dimensions = metadata.Dimensions;

dimensions[0] = 0;

long[] shape = dimensions.Select(x => (long)x).ToArray();
var ortValue = OrtValue.CreateAllocatedTensorValue(
OrtAllocator.DefaultInstance, metadata.ElementDataType, shape);
return ortValue;
}
```

Not sure if this is the correct way of handling a dynamic shape but the error disappeared.
That being said, most likely is incorrect because outputs returns empty.

I would like to kindly ask if, or when you adapt the code to handle dynamic shapes, could you confirm if it works with the .onnx file for the 6d pose estimation github example? https://github.com/TexasInstruments/edgeai-yolox/blob/main/demo/ONNXRuntime/onnx_inference_object_pose.py

The python code looks extremely familiar on the post process side so i can only assume i made a mistake by setting 0 as a value and that is not how dynamic shape is handled.

Thanks in advance!

**Comment by tv-gc** [comment]
Hi again!

I was reading the official documentation and, at least, for this particular case, i think i could get away with having the pre allocate a tensor for dynamic shape for the output and let onnxruntime handle it afterwards.

       ` var outputs = session.Run(new RunOptions(), session.InputNames, inputs, session.OutputNames);`

That being said, the result is still empty and i cannot understand why.

Is this still the incorrect way of handling this situation or am i missing something extra on top of it?

cheers!!

**Comment by asus4** [comment]
The above way looks correct to me.

But  `dimensions[0] = 0;` is incorrect. It must be more than 0.

How about setting the shape directly?

```c#
long[] shape = new[] {1, 15};
// or {2, 15}, {3, 15} 
var value = OrtValue.CreateAllocatedTensorValue(OrtAllocator.DefaultInstance, metadata.ElementDataType, shape);
```

(It may be wrong, as correct execution requires understanding the model's structure. And I haven't checked the model.)

**Comment by tv-gc** [comment]
I tried putting it at 1 but it gives this error

`OnnxRuntimeException: [ErrorCode:RuntimeException] Non-zero status code returned while running Gather node. Name:'/1/Gather_8' Status Message: D:\a\_work\1\s\onnxruntime\core\framework\execution_frame.cc:173 onnxruntime::IExecutionFrame::GetOrCreateNodeOutputMLValue shape && tensor.Shape() == *shape was false. OrtValue shape verification failed. Current shape:{1,15} Requested shape:{0,15}`

I spent the entire day trying to make it work but for this particular model, i can't do it.

**Comment by tv-gc** [comment]
I do believe the problem relies on the input image pre-processing stage but from the python example i don't see anything special about it

```
origin_img = cv2.imread(image_path)
origin_img = cv2.resize(origin_img, input_shape, interpolation=cv2.INTER_LINEAR)
img = origin_img.transpose((2, 0, 1))
img = np.ascontiguousarray(img, dtype=np.float32)
ort_inputs = {session.get_inputs()[0].name: img[None, :, :, :]}
```

Does the TextureToTensor do the same as this or is it necessary to create the input from texture from scratch?

**Comment by asus4** [comment]
@tv-gc 
The pre-processing will have the same result with the following option settings.

AspectMode: None
Mean: (0,0,0)
Std: (1,1,1)

![Screenshot 2024-02-12 at 10 26 27](https://github.com/asus4/onnxruntime-unity-examples/assets/357497/174bb0de-7ad2-4626-bc6c-29cb4dcf762c)

**Comment by tv-gc** [comment]
@asus4 Unfortunately it was not the problem and i don't know what else it may be.

The input name is correct (images), the output name is also correct (detections)
Outputs is handled by the session so it doesn't break with dynamic shape values.
It simply returns empty with this onnx model file.
As soon as i switch to another file, like the yolox_nano or other for regular object detection, the output result has values.

I am testing with a 640x480 image that should return results, i have also tried to set it as RGB24bit, but doesn't do anything.
(also tried with several other images just to confirm it was not this particular image's fault)

![ti_mustard](https://github.com/asus4/onnxruntime-unity-examples/assets/90689163/c6cb0802-1b2e-4082-8e0b-3d1bd518915b)

**Comment by asus4** [comment]
It looks like the ONNX Runtime issue to me.

To clarify the issue, please verify the following steps:

1.  the original code `onnx_inference_object_pose.py` works as intended.
2. Port it to the ONNX Runtime C# v1.17.0 (Not Unity), and it works as intended.

**Comment by tv-gc** [comment]
I have tested the original python code and it works.
I am now trying to port it to onnx runtime and see if it breaks.

I'll give more updates when done

**Comment by tv-gc** [comment]
Hi @asus4 

Sorry for the late reply

I have ported the code to native onnxruntime c# and it is working.
So, unfortunatly, it is somewhere in the Unity port that it is breaking.

Here's the native c# code i used to test the python version, along with the image. 

   ```
 public static void Main(string[] args)
    {
        // Read paths
        string modelFilePath = args[0];
        string imageFilePath = args[1];
        string outImageFilePath = args[2];

        // Read image
        using Image<Rgb24> image = Image.Load<Rgb24>(imageFilePath);

        // Resize image
        float ratio = 800f / Math.Min(image.Width, image.Height);
        float ratioHeight = 480f / image.Height;
        float ratioWidth = 640f / image.Width;
        image.Mutate(x => x.Resize((int)(ratioWidth * image.Width), (int)(ratioHeight * image.Height)));

        // Preprocess image
        var paddedHeight = (int)(Math.Ceiling(image.Height / 32f) * 32f);
        var paddedWidth = (int)(Math.Ceiling(image.Width / 32f) * 32f);
        DenseTensor<float> input = new DenseTensor<float>(new[] { 3, paddedHeight, paddedWidth });
        var mean = new[] { 1, 0, 1 };
        image.ProcessPixelRows(accessor =>
        {
            for (int y = paddedHeight - accessor.Height; y < accessor.Height; y++)
            {
                Span<Rgb24> pixelSpan = accessor.GetRowSpan(y);
                for (int x = paddedWidth - accessor.Width; x < accessor.Width; x++)
                {
                    input[0, y, x] = pixelSpan[x].B - mean[0];
                    input[1, y, x] = pixelSpan[x].G - mean[1];
                    input[2, y, x] = pixelSpan[x].R - mean[2];
                }
            }
        });

        var inputOrtValue = OrtValue.CreateTensorValueFromMemory(OrtMemoryInfo.DefaultInstance, input.Buffer, new long[] {1, 3, paddedHeight, paddedWidth});
        // Setup inputs and outputs
        var inputs = new Dictionary<string, OrtValue>
        {
            { "images", inputOrtValue }
        };

        // Run inference
        using var session = new InferenceSession(modelFilePath);
        using IDisposableReadOnlyCollection<OrtValue> results = session.Run(new RunOptions(), inputs, session.OutputNames);

        // Postprocess to get predictions
        var resultsArray = results.ToArray();

        Console.WriteLine(resultsArray.Length);

        var result = resultsArray[0].GetTensorDataAsSpan<float>();

        Console.WriteLine(result.Length);

        var score = result[4];

        Console.WriteLine(score);

        //long[] labels = resultsArray[1].AsEnumerable<long>().ToArray();
        //float[] confidences = resultsArray[2].AsEnumerable<float>().ToArray();
        var predictions = new List<Prediction>();
        //var minConfidence = 0.7f;
        //for (int i = 0; i < boxes.Length - 4; i += 4)
        //{
        //    var index = i / 4;
        //    if (confidences[index] >= minConfidence)
        //    {
        //        predictions.Add(new Prediction
        //        {
        //            Box = new Box(boxes[i], boxes[i + 1], boxes[i + 2], boxes[i + 3]),
        //            Label = LabelMap.Labels[labels[index]],
        //            Confidence = confidences[index]
        //        });
        //    }
        //}

        predictions.Add(new Prediction
        {
            Box = new Box(result[0], result[1], result[2], result[3]),
            Label = "test",
            //Label = LabelMap.Labels[labels[index]],
            Confidence = score
        });

        // Put boxes, labels and confidence on image and save for viewing
        using var outputImage = File.OpenWrite(outImageFilePath);
        Font font = SystemFonts.CreateFont("Arial", 16);
        foreach (var p in predictions)
        {
            image.Mutate(x =>
            {
                x.DrawLines(Color.Red, 2f, new PointF[] {

                    new PointF(p.Box.Xmin, p.Box.Ymin),
                    new PointF(p.Box.Xmax, p.Box.Ymin),

                    new PointF(p.Box.Xmax, p.Box.Ymin),
                    new PointF(p.Box.Xmax, p.Box.Ymax),

                    new PointF(p.Box.Xmax, p.Box.Ymax),
                    new PointF(p.Box.Xmin, p.Box.Ymax),

                    new PointF(p.Box.Xmin, p.Box.Ymax),
                    new PointF(p.Box.Xmin, p.Box.Ymin)
                });
                x.DrawText($"{p.Label}, {p.Confidence:0.00}", font, Color.White, new PointF(p.Box.Xmin, p.Box.Ymin));
            });
        }
        image.SaveAsJpeg(outputImage);
    }
}
```

This is basically this code https://onnxruntime.ai/docs/tutorials/csharp/fasterrcnn_csharp.html adapted to run the yolo_s_object_pose.onnx file.

And the image
![demo](https://github.com/asus4/onnxruntime-unity-examples/assets/90689163/30cf3686-1cf9-47c1-8d1b-b3422ff18488)

**Comment by asus4** [comment]
@tv-gc can you share your Unity code?  The input image needs to be BGR, not RGB and range of 0 ~ 255?

**Comment by tv-gc** [comment]
Of course, here's the package 
I am running on the Yolox scene, most of it is as the default code you made, just changed some of the inference code for testing and using a texture instead of webcam for faster iterations of testing.

[Yolox.zip](https://github.com/asus4/onnxruntime-unity-examples/files/14350226/Yolox.zip)
[InferenceDynamic.zip](https://github.com/asus4/onnxruntime-unity-examples/files/14350236/InferenceDynamic.zip)

**Comment by asus4** [comment]
@tv-gc thanks. After I looked into your code, I found that the image pre-processing was not correct. The input image requires BGR, 0 ~ 255 range in this model, which is not supported in TextureToTensor.cs

You will get the output once the input tensor is properly provided.

![Screenshot 2024-02-20 at 21 49 40](https://github.com/asus4/onnxruntime-unity-examples/assets/357497/e530ec48-4dcc-4f22-9872-c6d876bae53f)

here's the minimum test code,

```c#
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using Microsoft.ML.OnnxRuntime.Unity;
using UnityEngine;

public class Yolo6DPoseSample : MonoBehaviour
{
    [SerializeField]
    TextAsset model;

    [SerializeField]
    Texture2D inputTexture;

    [SerializeField]
    RectTransform previewRect;

    float[,] results;

    void Start()
    {
        // Make session
        using var sessionOptions = new SessionOptions();
        using var runOptions = new RunOptions();
        using var session = new InferenceSession(model.bytes, sessionOptions);

        session.LogIOInfo();

        // Preprocessing

        var input = new DenseTensor<float>(new[] { 3, 480, 640 });
        Color32[] pixels = inputTexture.GetPixels32();
        for (int i = 0; i < pixels.Length; i++)
        {
            int x = i % inputTexture.width;
            int y = i / inputTexture.width;
            y = inputTexture.height - y - 1;
            input[0, y, x] = pixels[i].b;
            input[1, y, x] = pixels[i].g;
            input[2, y, x] = pixels[i].r;
        }
        var inputs = new OrtValue[]
        {
            OrtValue.CreateTensorValueFromMemory(OrtMemoryInfo.DefaultInstance,
                input.Buffer, new long[] {1, 3, 480, 640})
        };


        // Run inference
        using var outputs = session.Run(runOptions, session.InputNames, inputs, session.OutputNames);

        // Postprocess
        var info = outputs[0].GetTensorTypeAndShape();
        var shape = info.Shape;
        Debug.Log($"out info: elements: {info.ElementCount}, shape: {string.Join(",", shape)}");
        var result = outputs[0].GetTensorDataAsSpan<float>();
        results = new float[shape[0], shape[1]];
        for (int i = 0; i < shape[0]; i++)
        {
            for (int j = 0; j < shape[1]; j++)
            {
                results[i, j] = result[i * (int)shape[1] + j];
            }
        }

        // Cleanup
        foreach (var inputValue in inputs)
        {
            inputValue.Dispose();
        }
    }

    void Update()
    {
        DrawBoxes();
    }

    private void DrawBoxes()
    {
        if (results == null)
        {
            return;
        }

        var mtx = previewRect.localToWorldMatrix;
        const int H_WIDTH = 640 / 2;
        const int H_HEIGHT = 480 / 2;
        Color color = Color.green;

        for (int i = 0; i < results.GetLength(0); i++)
        {
            float x1 = results[i, 0] - H_WIDTH;
            float y1 = -results[i, 1] + H_HEIGHT;
            float x2 = results[i, 2] - H_WIDTH;
            float y2 = -results[i, 3] + H_HEIGHT;

            Vector3 p1 = mtx.MultiplyPoint3x4(new Vector3(x1, y1, 0));
            Vector3 p2 = mtx.MultiplyPoint3x4(new Vector3(x1, y2, 0));
            Vector2 p3 = mtx.MultiplyPoint3x4(new Vector3(x2, y2, 0));
            Vector2 p4 = mtx.MultiplyPoint3x4(new Vector3(x2, y1, 0));
            Debug.DrawLine(p1, p2, color);
            Debug.DrawLine(p2, p3, color);
            Debug.DrawLine(p3, p4, color);
            Debug.DrawLine(p4, p1, color);
        }
    }
}

```

**Comment by asus4** [comment]
I will move this to discussions since this is not the issue of ONNXRuntime.
